firmware-management/mbl-watchdog-init/mbl-watchdog-init.c

```diff
@@ -0,0 +1,173 @@
+}
+
+
+int print_last_boot_reason(const int boot_status)
+{
+    switch (boot_status)
+    {
+        case WDIOF_OVERHEAT:
+            log_info("The last reboot was caused by the CPU overheating.");
+            return 0;
+        case WDIOF_CARDRESET:
+            log_info("The last reboot was caused by a watchdog reset.");
+            return 0;
+        case WDIOF_FANFAULT:
+            log_info("The last reboot was because a system fan monitored by the watchdog card failed.");
+            return 0;
+        case WDIOF_EXTERN1:
+            log_info("The last reboot was because external monitoring relay/source 1 was triggered.");
+            return 0;
+        case WDIOF_EXTERN2:
+            log_info("The last reboot was because external monitoring relay/source 2 was triggered.");
+            return 0;
+        case WDIOF_POWERUNDER:
+            log_info("The last reboot was due to the machine showing an undervoltage status.");
+            return 0;
+        case WDIOF_POWEROVER:
+            log_info("The last reboot was due to the machine showing an overvoltage status.");
+            return 0;
+    }
+
+    return -1;
+}
+
+
+int get_last_boot_status(const int watchdog_fd)
+{
+    int flags;
+
+    const int gs_ret = ioctl(watchdog_fd, WDIOC_GETBOOTSTATUS, &flags);
+    if (is_err(gs_ret))
+    {
+        return -1;
+    }
+
+    const int lb_ret = print_last_boot_reason(flags);
+    if (is_err(lb_ret))
+    {
+        log_info("Last boot wasn't caused by the watchdog card");
+    }
+
+    return 0;
+}
+
+//============================================================================
+// Main entry point
+int main(int argc, char **argv)
+{
+    int timeout = 0;
+    char* wdog_filename = "";
+
+    // Parse command line arguments
+    int current_opt;
+    int optindex;
+    struct option longopts[] = {
+        {"timeout", required_argument, NULL, 't'},
+        {"device", required_argument, NULL, 'w'},
+    };
+
+    while ((current_opt = getopt_long(argc, argv, "tw:", longopts, &optindex)) != -1)
+    {
+        switch (current_opt)
+        {
+            case 't':
+                timeout = atoi(optarg);
+                break;
+            case 'w':
+                wdog_filename = optarg;
+                break;
+            default:
+                log_warning("Only --timeout and --wdog-fname should be provided. Exiting.");
+                return EXIT_FAILURE;
+        }
+    }
+
+    const int wdog_fd = open(wdog_filename, O_WRONLY);
+    if (is_err(wdog_fd))
+    {
+        return log_error("Failed to open watchdog device file");
+    }
+
+    const int gs_ret = get_last_boot_status(wdog_fd);
+    if (is_err(gs_ret))
+    {
+        return log_error("Failed to get the last boot status");
+    }
+
+    const int to_ret = set_watchdog_timeout(wdog_fd, &timeout);
+    if (is_err(to_ret))
+    {
+        return log_error("Failed to set the watchdog timeout");
+    }
+
+    const int cl_ret = close(wdog_fd);
```

Comment:
This won't get called if any of the `ioctl`s fail. Unfortunately C doesn't have `finally` or C++ style RAII to make sure things are cleaned up automatically... You might want to use `goto` to simulate `finally` as in https://eli.thegreenplace.net/2009/04/27/using-goto-for-error-handling-in-c.

I suppose you could argue that Linux will just clean up after you anyway, but if you do that I'd want to see a big comment explaining why you don't always close your fd because it would make this code unsafe to be re-used in a longer-lived program.

Reply:
131dc3fca138422be97e2ca07dbcca09f5c2cd14